pkg/referrerstore/oras/authprovider/azure/azureworkloadidentity.go

```diff
@@ -0,0 +1,168 @@
+/*
+Copyright The Ratify Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azure
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/AzureAD/microsoft-authentication-library-for-go/apps/confidential"
+	"github.com/containerd/containerd/reference"
+	provider "github.com/deislabs/ratify/pkg/referrerstore/oras/authprovider"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"k8s.io/klog/v2"
+)
+
+type AzureWIProviderFactory struct{}
+type azureWIAuthProvider struct{}
+
+type azureWIAuthProviderConf struct {
+	Name string `json:"name"`
+}
+
+const (
+	azureWIAuthProviderName      string = "azure-wi"
+	dockerTokenLoginUsernameGUID string = "00000000-0000-0000-0000-000000000000"
+)
+
+// init calls Register for our Azure Workload Identity provider
+func init() {
+	provider.Register(azureWIAuthProviderName, &AzureWIProviderFactory{})
+	logrus.Info("Azure-WI provider registered")
+}
+
+// Create returns an AzureWIAuthProvider
+func (s *AzureWIProviderFactory) Create(authProviderConfig provider.AuthProviderConfig) (provider.AuthProvider, error) {
+	conf := azureWIAuthProviderConf{}
+	authProviderConfigBytes, err := json.Marshal(authProviderConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := json.Unmarshal(authProviderConfigBytes, &conf); err != nil {
+		return nil, fmt.Errorf("failed to parse auth provider configuration: %v", err)
+	}
+	logrus.Info("Azure-WI provider created succesfully")
+	return &azureWIAuthProvider{}, nil
+}
+
+// Enabled always returns true since there are no fields to verify
+func (d *azureWIAuthProvider) Enabled() bool {
+	return true
+}
+
+// Provide returns the credentials for a specificed artifact.
+// Uses Azure Workload Identity to retrieve an AAD access token which can be
+// exchanged for a valid ACR refresh token for login.
+func (d *azureWIAuthProvider) Provide(artifact string) (provider.AuthConfig, error) {
+	tenantID := os.Getenv("AZURE_TENANT_ID")
+
+	// parse the artifact reference string to extract the registry host name
+	parsedSpec, err := reference.Parse(artifact)
+	if err != nil {
+		return provider.AuthConfig{}, err
+	}
+
+	artifactHostName := parsedSpec.Hostname()
+
+	// retrieve an AAD Access token from ARM
+	aadToken, err := getAADAccessToken(tenantID, "https://management.azure.com/")
```

Comment:
ACR is moving away from arm audienced AAD token, can can pass the ACR audience instead https://containerregistry.azure.net